source/validate.cpp

```diff
@@ -520,24 +538,117 @@ spv_result_t ModuleLayoutPass(ValidationState_t& _,
               inst->words + inst->operands[2].offset;
           if (*storage_class != SpvStorageClassFunction)
             return _.diag(SPV_ERROR_INVALID_LAYOUT)
-                   << "All OpVariable instructions in a function must have a "
-                      "storage class of Function[7]";
+                   << "All Variable instructions in a function must have a "
+                      "storage class of function[7]";
+        } break;
+        default:
           break;
+      }
+      if (_.getLayoutStage() == kLayoutFunctionDeclarations) {
+        switch (opcode) {
+          case SpvOpFunction:
+            if (_.in_function_body()) {
+              return _.diag(SPV_ERROR_INVALID_LAYOUT)
+                     << "Cannot declare a function in a function body";
+            }
+            spvCheckReturn(_.get_functions().RegisterFunction(
+                inst->result_id, inst->type_id,
+                inst->words[inst->operands[2].offset],
+                inst->words[inst->operands[3].offset]));
+            break;
+          case SpvOpFunctionParameter:
+            if (_.in_function_body() == false) {
+              return _.diag(SPV_ERROR_INVALID_LAYOUT) << "Function parameter "
+                                                         "instructions must be "
+                                                         "in a function body";
+            }
+            spvCheckReturn(_.get_functions().RegisterFunctionParameter(
```

Comment:
This is now checked in the RegisterFunctionParameter function. See [implementation](https://github.com/KhronosGroup/SPIRV-Tools/pull/51/files#diff-84a5aa9a556fde9600c6f5583e34ed81R253) and [unit tests](https://github.com/KhronosGroup/SPIRV-Tools/pull/51/files#diff-6dc2086e9da925d662089f2876b41bb7R380).